spacy/tokens/doc.pyx

```diff
@@ -1462,6 +1462,152 @@ cdef class Doc:
                 remove_label_if_necessary(attributes[i])
                 retokenizer.merge(span, attributes[i])
 
+    def from_json(self, doc_json):
+        """Convert a JSON document generated by Doc.to_json() to a Doc.
+
+        vocab (Vocab): A storage container for lexical types.
+        RETURNS (Doc): A doc instance corresponding to the specified JSON representation.
+        """
+
+        ### Token-level properties ###
+
+        words = []
+        # Map annotation type IDs to their string equivalents.
+        annot_types = {POS: "pos", HEAD: "head", DEP: "dep", LEMMA: "lemma", TAG: "tag", MORPH: "morph"}
+        token_annotations = {
+            # For each annotation type: store (1) annotation type required to include it, (2) list of annotations for
+            # this type.
+            annot_type_id: {"req": annot_type_id if annot_type_id != HEAD else DEP, "values": None}
+            for annot_type_id in annot_types
+        }
```

Comment:
I think the code will be more readable if this is split into two separate dicts, so that we don't have `["values"]` everywhere below.